spec/ValidationAndPasswordsReset.spec.js

```diff
@@ -910,6 +910,113 @@ describe('Custom Pages, Email Verification, Password Reset', () => {
     });
   });
 
+  it('should programmatically reset password on ajax request', done => {
+    const user = new Parse.User();
+    const emailAdapter = {
+      sendVerificationEmail: () => Promise.resolve(),
+      sendPasswordResetEmail: options => {
+        request({
+          url: options.link,
+          followRedirects: false,
+        }).then(response => {
+          expect(response.status).toEqual(302);
+          const re = /http:\/\/localhost:8378\/1\/apps\/choose_password\?token=([a-zA-Z0-9]+)\&id=test\&username=zxcv/;
+          const match = response.text.match(re);
+          if (!match) {
+            fail('should have a token');
+            done();
+            return;
+          }
+          const token = match[1];
+
+          request({
+            url: 'http://localhost:8378/1/apps/test/request_password_reset',
+            method: 'POST',
+            body: { new_password: 'hello', token, username: 'zxcv' },
+            headers: {
+              'Content-Type': 'application/x-www-form-urlencoded',
+              'X-Requested-With': 'XMLHttpRequest',
+            },
+            followRedirects: false,
+          }).then(response => {
+            console.log('REQUEST RESPONSE');
+            expect(response.status).toEqual(200);
+            expect(response.text).toEqual('"Password successfully reset"');
+
+            Parse.User.logIn('zxcv', 'hello').then(
+              function() {
+                const config = Config.get('test');
+                config.database.adapter
+                  .find(
+                    '_User',
+                    { fields: {} },
+                    { username: 'zxcv' },
+                    { limit: 1 }
+                  )
+                  .then(results => {
+                    // _perishable_token should be unset after reset password
+                    expect(results.length).toEqual(1);
+                    expect(results[0]['_perishable_token']).toEqual(undefined);
```

Comment:
Can you add a check that `_perishable_token` was set before this?

Reply:
Why is it needed here? All we need is to make sure user does not stay logged in after resetting his password.